mass/core/query_handler.py

```diff
@@ -50,16 +53,35 @@ async def handle_query(
         filters: list[models.Filter],
         skip: int = 0,
         limit: Optional[int] = None,
-    ) -> list[models.Resource]:
+    ) -> models.QueryResults:
         """Return resources that match query"""
+
+        # get configured facet fields for given resource class
+        try:
+            facet_fields = self._config.searchable_classes[
+                class_name
+            ].facetable_properties
+        except KeyError as err:
+            raise ClassNotConfiguredError(class_name=class_name) from err
+
+        # build the aggregation pipeline
         pipeline = utils.build_pipeline(
-            query=query, filters=filters, skip=skip, limit=limit
+            query=query,
+            filters=filters,
+            facet_fields=facet_fields,
+            skip=skip,
+            limit=limit,
         )
+
+        # run the aggregation. Results will have {facets, count, hits} format
         aggregator = self._aggregator_collection.get_aggregator(class_name=class_name)
         aggregator_results: list[JsonObject] = await aggregator.aggregate(
             pipeline=pipeline
         )
-        query_results: list[models.Resource] = [
-            utils.document_to_resource(document=item) for item in aggregator_results
-        ]
+        query_results = models.QueryResults(**aggregator_results[0])
```

Comment:
What is actually the result from the aggregator? Why is only the first element used?

Reply:
Yes, that would be great. Otherwise the internal implementation details of the aggregator are leaked to the outside.

Reply:
Fixed as part of [8c8d42c](https://github.com/ghga-de/mass/pull/5/commits/8c8d42cb32e9b4c13a55d49e53e747d6588e599e)